src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs

```diff
@@ -404,17 +404,27 @@ private async Task ReadMagicAsync(byte[] magic, CancellationToken cancellation)
         /// <exception cref="OperationCanceledException">Thrown if the operation was cancelled or the end of the stream was reached.</exception>
         private async Task ReadBytesAsync(byte[] buffer, int offset, int bytesToRead, CancellationToken cancellation = default(CancellationToken))
         {
-            while (bytesToRead > 0)
+            using (await this.writeLock.LockAsync(cancellation).ConfigureAwait(false))
             {
-                int chunkSize = await this.stream.ReadAsync(buffer, offset, bytesToRead, cancellation).ConfigureAwait(false);
-                if (chunkSize == 0)
+                if (this.stream == null)
```

Comment:
Perhaps the fix here is to just check that the stream isn't null?

Reply:
Wouldn't fix the issue, even if we check for null, another thread could set it to null before we consume it...

Reply:
The fix is to take a local scope of the stream and use that, then another thread cannot null it

Reply:
I don't think this is the right fix, the write lock is to prevent multiple writes o the stream, while when reading we only have a single thread that reads from it, and we can just easy bypass by doing a  null check